### PR TITLE
[#133] feat(netty): Add StreamServer.

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?

Add StreamServer for netty replace grpc.

### Why are the changes needed?
Add StreamServer.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

UT.

### CODE REVIEW

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 1]
```diff
@@ -404,6 +404,48 @@ public class ShuffleServerConf extends RssBaseConf {
       .defaultValue(-1)
       .withDescription("Shuffle netty server port");
 
+  public static final ConfigOption<Boolean> NETTY_SERVER_ENABLED = ConfigOptions
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:407]: We don't need this config option. If netty port is set, we should enable Netty server.

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 2]
```diff
@@ -404,6 +404,48 @@ public class ShuffleServerConf extends RssBaseConf {
...
+      .key("rss.server.netty.enable")
+      .booleanType()
+      .defaultValue(false)
+      .withDescription("If enable netty server");
+
+  public static final ConfigOption<Boolean> SERVER_UPLOAD_EPOLL_ENABLE = ConfigOptions
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:413]: What's the meaning of `upload`?

File: server/src/main/java/org/apache/uniffle/server/netty/decoder/StreamServerInitDecoder.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.uniffle.server.ShuffleServer;
+
+public class StreamServerInitDecoder extends ByteToMessageDecoder {
+
+  private static final Logger logger = LoggerFactory.getLogger(StreamServerInitDecoder.class);
+
+  private ShuffleServer shuffleServer;
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/netty/decoder/StreamServerInitDecoder.java:34]: Why do we need `ShuffleServer`?

**Comment by jerqi** [comment]
@smallzhongfeng Could you help me review this pr?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java (hunk 2 above)
Comment (leixm) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java]: Already Fixed.

File: server/src/main/java/org/apache/uniffle/server/netty/decoder/StreamServerInitDecoder.java (hunk 3 above)
Comment (leixm) [reply server/src/main/java/org/apache/uniffle/server/netty/decoder/StreamServerInitDecoder.java]: In fact, StreamServerInitDecoder needs ShuffleServer to obtain taskManager, etc. I will delete it in this PR, and add it to subsequent PRs.

File: server/src/main/java/org/apache/uniffle/server/netty/StreamServer.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                                          : new ServerBootstrap().group(bossGroup, workerGroup)
+                                                .channel(NioServerSocketChannel.class);
+
+    return serverBootstrap.childHandler(new ChannelInitializer<SocketChannel>() {
+      @Override
+      public void initChannel(final SocketChannel ch) {
+        ch.pipeline().addLast(handlerSupplier.get());
+      }
+    })
+               .option(ChannelOption.SO_BACKLOG, backlogSize)
```
Comment (smallzhongfeng) [inline server/src/main/java/org/apache/uniffle/server/netty/StreamServer.java:86]: Maybe we should add Option `TCP_NODELAY`, this parameter should be used to send packets this time. It does not need to use Nagle algorithm to reduce the delay time.

File: server/src/main/java/org/apache/uniffle/server/netty/StreamServer.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      ExitUtils.terminate(1, "Fail to start stream server", e, LOG);
+    }
+  }
+
+  public void stop() {
+    if (channelFuture != null) {
+      channelFuture.channel().close().awaitUninterruptibly(10L, TimeUnit.SECONDS);
+      channelFuture = null;
+    }
+    shuffleBossGroup.shutdownGracefully();
```
Comment (smallzhongfeng) [inline server/src/main/java/org/apache/uniffle/server/netty/StreamServer.java:119]: `ShuffleBossGroup` determines whether it is empty ?

File: server/src/main/java/org/apache/uniffle/server/netty/decoder/StreamServerInitDecoder.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    if (in.readableBytes() < Byte.BYTES) {
+      return;
+    }
+    in.markReaderIndex();
+    byte magicByte = in.readByte();
+    in.resetReaderIndex();
+
+    addDecoder(ctx, magicByte);
+  }
+
```
Comment (smallzhongfeng) [inline server/src/main/java/org/apache/uniffle/server/netty/decoder/StreamServerInitDecoder.java:48]: Remove this blank line.

**Comment by leixm** [comment]
cc@smallzhongfeng  All done.

**Comment by leixm** [comment]
@smallzhongfeng  All done.

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 7]
```diff
@@ -404,6 +404,58 @@ public class ShuffleServerConf extends RssBaseConf {
       .defaultValue(-1)
       .withDescription("Shuffle netty server port");
 
+  public static final ConfigOption<Boolean> NETTY_SERVER_EPOLL_ENABLE = ConfigOptions
+      .key("rss.server.netty.epoll.enable")
+      .booleanType()
+      .defaultValue(false)
+      .withDescription("If enable epoll model with netty server");
```
[the remark this answers is not in the document]
Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:411]: We set up an issue first.

[the remark this answers is not in the document]
Comment (leixm) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:411]: EPOLL is more suitable for scenarios with a large number of connections, but we need more tests to verify the difference between nio and epoll in uniffle usage scenarios. It is difficult to give a suggestion at present, so the default value is nio.

File: server/src/main/java/org/apache/uniffle/server/ShuffleServer.java [hunk 8]
```diff
@@ -221,6 +230,10 @@ private void initialization() throws Exception {
     shuffleBufferManager = new ShuffleBufferManager(shuffleServerConf, shuffleFlushManager);
     shuffleTaskManager = new ShuffleTaskManager(shuffleServerConf, shuffleFlushManager,
         shuffleBufferManager, storageManager);
+    nettyServerEnabled = shuffleServerConf.get(ShuffleServerConf.NETTY_SERVER_PORT) > 0;
```
[the remark this answers is not in the document]
Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServer.java]: We set up an issue first.

**Review by smallzhongfeng — APPROVED** [review]
LGTM

File: server/src/main/java/org/apache/uniffle/server/netty/StreamServer.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamServer.class);
+
+  private ShuffleServer shuffleServer;
+  private EventLoopGroup shuffleBossGroup;
+  private EventLoopGroup shuffleWorkerGroup;
+  private ShuffleServerConf shuffleServerConf;
+  private ChannelFuture channelFuture;
+
+  public StreamServer(ShuffleServer shuffleServer) {
```
[the remark this answers is not in the document]
Comment (leixm) [reply server/src/main/java/org/apache/uniffle/server/netty/StreamServer.java:53]: sure. I will add in later PRs.

**Comment by leixm** [comment]
@advancedxy  All done.

**Comment by leixm** [comment]
@jerqi  @advancedxy  Can you help review plz?

**Review by jerqi — APPROVED** [review]
LGTM, thanks @advancedxy @leixm @smallzhongfeng  merged.

**Pull request merged**